kcidb/io/schema/v3.py

```diff
@@ -0,0 +1,626 @@
+SHA1_PATTERN = "[0-9a-f]{40}"
+
+# A regular expression pattern matching strings containing sha256 hashes
+SHA256_PATTERN = "[0-9a-f]{64}"
+
+# A regular expression pattern matching strings containing Git repository
+# commit hash (sha1 or sha256)
+GIT_REPOSITORY_COMMIT_HASH_PATTERN = f"({SHA1_PATTERN}|{SHA256_PATTERN})"
+
+# A regular expression pattern matching strings containing revision IDs
+REVISION_ID_PATTERN = \
+    f"{GIT_REPOSITORY_URL_PATTERN}" \
+    f"@{GIT_REPOSITORY_COMMIT_HASH_PATTERN}" \
+    f"(\\+{SHA256_PATTERN})?"
+
+# A regular expression pattern matching strings containing origin name
+ORIGIN_PATTERN = "[a-z0-9_]+"
+
+# A regular expression pattern matching strings containing an object ID
+ORIGIN_ID_PATTERN = f"{ORIGIN_PATTERN}:.*"
+
+# JSON schema for a named remote resource
+JSON_RESOURCE = {
+    "title": "resource",
+    "description": "A named remote resource",
+    "type": "object",
+    "properties": {
+        "name": {
+            "type": "string",
+            "description":
+                "Resource name. Must be usable as a local file name for the "
+                "downloaded resource. Cannot be empty. Should not include "
+                "directories.",
+            "pattern": "^[^/]+$",
+        },
+        "url": {
+            "type": "string",
+            "format": "uri",
+            "description":
+                "Resource URL. Must point to the resource file directly, "
+                "so it could be downloaded automatically.",
+        },
+    },
+    "additionalProperties": False,
+    "required": [
+        "name",
+        "url",
+    ],
+    "examples": [
+        {
+            "name": "console.log",
+            "url":
+                "https://artifacts.cki-project.org/pipelines/223563/logs/"
+                "aarch64_host_1_console.log"
+        },
+        {
+            "name": "kernel.tar.gz",
+            "url":
+                "https://artifacts.cki-project.org/pipelines/224569/"
+                "kernel-stable-aarch64-"
+                "a2fc8ee6676067f27d2f5c6e4d512adff3d9938c.tar.gz"
+        }
+    ]
+}
+
+# JSON schema for a code revision
+JSON_REVISION = {
+    "title": "revision",
+    "description":
+        "A revision of the tested code.\n"
+        "\n"
+        "Represents a way the tested source code could be obtained. E.g. "
+        "checking out a particular commit from a git repo, and applying a "
+        "set of patches on top.",
+    "type": "object",
+    "properties": {
+        "id": {
+            "type": "string",
+            "description":
+                "Revision ID.\n"
+                "\n"
+                "Must be a string containing the following, in order.\n"
+                "\n"
+                "The URL of the Git repository which contains the base code "
+                "of the revision. The shortest possible https:// URL, or, if "
+                "that's not available, the shortest possible git:// URL."
```

Comment:
I think this can be a bit fragile, as trees move from time to time.  In the same way that we have a catalogue of test suites, having a catalogue of trees with simple aliases to be able to refer to them indirectly would avoid this issue, and also make it easier to understand for humans.

See the list of trees in [`build-configs.yaml`](https://github.com/kernelci/kernelci-core/blob/master/build-configs.yaml#L5).

Reply:
I agree, we won't handle renames this way, but then the impact of losing tree continuity doesn't seem too high right now, while complicating the submission process. I think the time for such a feature is probably when we got at least a few submitting CI systems and they all feel this is needed.

Reply:
At the same time, we *can* have the alias table for use in report WebUI to shorten those URLs.

Reply:
Agreed.  We've had some trees move several times, and keeping the same "short name" (aka tree-name) with a changing URL has been very useful for the users.

I find the full URL in the current grafana UI cumbersome, and hard to scan visually (it also takes up too much screen real estate.)

Reply:
@khilman Do you agree with using tree names at submission time or the web UI based approach (which doesn't prevent against moving trees)?

Reply:
We can have one alias per multiple repo URLs to handle moving trees in the UI.
I want to avoid forcing all the CI systems to consult the mapping before they can push the data.

Reply:
I think tree-names at submission time makes the most sense.  

To not overburden new CI systems, that don't know about tree-names, kcidb can manage the look-up table (ideally using existing `build-configs.yaml`).  URLs that don't have an entry in the table, can be flagged/fixed.

Reply:
@khilman, I respectfully disagree. Tree names are subjective and debatable. Repo URLs are objective and straight-forward. They're simply places where you got the commit from. You **have** to specify them to Git to get your code to test. They are a hard reality. You don't have to know the tree name **at all** to get the code, build it, and run tests. They're simply a sugar coating. Having them in the data itself instead of URLs **will** complicate things.

So far we managed to avoid the need for one-way data transformation when submitting. Right now what you submit is **exactly** what you get back (if not, it should be fixed). This simplifies thinking about the system greatly. Doing a fixup on repo URLs at submission time will make stored and returned data different from submitted data. If we also put tree names into IDs, we'll have to fixup them too (both in revisions and in builds), further complicating the system. And if we don't, repo moving still won't be handled.

If we allow data transformation on submission there will be no going back. We still have a lot of problems to solve and we could do without having to think about **that**.

What we can do is have a URL->name table in the same dataset, provide command-line tools for modifying it and/or syncing it with build-configs.yaml, outside the normal report submission process (e.g. for administrators), and use that table when displaying the WebUI, and generating reports. We could use that table when making queries as well. I think I could implement that relatively quickly and we would have short tree names in the WebUI and in reports. That would handle repo moving too.

However, I think there's no place for tree names in the report data itself.

Reply:
BTW, we could put the tests.yaml into the dataset at the same time, as it would be similar, and then we could raise warnings on submission if a test isn't found. Same with repo URLs: we could raise warnings if a repo URL has no tree name assigned to it on submission.

Reply:
No need to transform the data, just submit both a URL and a tree name.  You can get that back exactly as well.

Reply:
@gctucker, I was thinking about that earlier, and gave it more thought now. This would work, provided we don't use the tree name as an identifier, but just as another property of the revision to display, or e.g. to filter by. Like this:

![image](https://user-images.githubusercontent.com/549246/82802553-c4568700-9e87-11ea-8f2f-a0f8eeec129e.png)

We would have to accept that not everyone will be able to submit those (at least at first), and that incorrect tree names will be submitted. We can provide support for making sure the tree names are alright, though (warning/erroring on submission, if requested).

Reply:
Yes, having both is what I meant, and the name is "just info".  The definitive ID should still be the URL, and provide some warnings/errors if a known name is not matching a known URL.

Reply:
Seems like a good compromise between using solely the URLs and enforcing short tree names.

Thinking about it, somewhere in the system, there will need to be some translation mechanism between what "native" data is available on the client side and what kcidb expects.  Test names, tree names, device names, kernel revision names...  We'll need to decide if we leave that up to each client to comply and do their own translation, or if the submitting tool can help by doing some of that automatically for deterministic data.  It seems to me that upstream tree URLs and the short names fall into the deterministic category.

Reply:
@gctucker, yes, this would work for now.

I'm still of the opinion that it's better to submitt pure hard data that needs no transformation, and has no extra data, and then have "dictionaries" translating that data in UI (such as looking up tree names based on repo URLs). However, let's see how this works out and then move towards what makes more sense.

I assume we don't have any more argument about changes in *this* PR and will merge it this evening unless somebody stops me. I will submit a separate PR adding the optional "tree_name" field to the schema next (I have the changes ready locally already).

Reply:
Yes it's up to you to decide when to merge this PR, I guess it's fine with a follow-up to add the tree name fields.

Because data from random sources will be in some unknown format, there is always going to be the need for some translation to happen before data gets saved in the database. Say, CI systems have different names for a same test suite, it will need to get translated so we don't have `LTP`, `ltp`, `Linux Testing Project`, `Ltp`... but only one name to rule them all.  Pushing the "hard data that needs no transformation" doesn't work there.  In the same way, if some CI systems use a `git://` URLs and others `https://` for a same tree, some translation will need to be done somewhere.  It pretty much applies to all the data being sent to a common database, by nature if there are no ways to enforce what these things should be in the original CI systems then they'll all tend to be different and will have to be transformed.

Reply:
@gctucker, yes there's that problem.

For now I'm hoping we'll be able to prevent that by giving the submitters incentives to submit complying data through better correlation and happier developers, and also by denying to submit whatever non-compliant data we can detect, ramping up the latter as we figure out more what we need. I.e. check data we receive is correct, do not admit incorrect data and let the submitters do fixes/mapping on their side.

I want to prevent, or to at least delay submission transformation as much as possible, as it will make the whole system much more obscure and harder to think about. I would rather free up my mind for thinking about more fundamental problems right now, like correlating revisions, describing, identifying and correlating test environments, presenting reports, notifications, and so on. I think we can refine our data later, one way or another.

Reply:
> For now I'm hoping we'll be able to prevent that by giving the submitters incentives to submit complying data

Exactly, and that means having the required translations done on the client side.  We can't remove this problem, unless we adopt a more integrated pipeline mechanism where test labs use the KernelCI builds and definitions with tree names and URLs, test suite names, device names etc... which is out of scope for the moment.

Some translations will _have_ to be done either:
- by the submitter before calling the provided tool to submit results (e.g. before kcidb-submit),
- by the submitting tool itself (e.g. by kcidb-submit),
- upon reception of the results on the server side (e.g. in Google Could Functions),
- or after getting it back from the database (e.g. dashboard and emails).

Where the translation happens and how is a problem that needs to be addressed, and in fact I believe it's the main thing we need to be able to provide a solution for in the context of a common database.  Starting with putting all this on the submitting end is probably easier initially for us but it won't make it more attractive for new CI systems to join.

Reply:
Yes, this has to be done somewhere.

I'm primarily against doing it ourselves on submission, client- or server-side. We'll have to augment/prettify some of the data on reporting side (dashboard/emails), yes, but I think it's reasonable to have the submitters do the mapping specific to their CI system, and it's too much to expect that from us, it simply won't scale. However, we can help them by picking which data is necessary to submit and in which shape.

Finally, I think our discussion is getting meta, too abstract, and we might be talking about different pieces of data and having different things in mind. Let's discuss the specific pieces as we get to them instead.

Reply:
In practice, for this particular case, I think it would make a lot of sense for clients to submit data with plain git URLs and the submitting tool would look up a tree name for that URL.

Reply:
@gctucker, if we can look up a tree name automatically based on the URL, why do we need to store it with each revision at all, why not just look it up when reporting?

Reply:
Because if the tree name is in the database, then you can directly do a query using it rather than "url 1" or "url 2" or "url 3" etc...  We're back to the 4 bullet points a couple of comments ago, I'm suggesting option 2 and you're suggesting option 4.  It's going to be a recurring theme, I guess semantics and the scope where data is meaningful should help us decide where these translations need to happen ideally.  Then some practical limitations may also come into play.

Reply:
We can have the tree names in a separate table, as I proposed earlier, and do a join if we need to query by a tree name. It's not the simplest-possible query, as we would have with the tree name right there in the revision, but it's still basic and is what databases are built for.

Regardless of that, I wonder why would you prefer doing the transformation at submission time (apart from making it easier for submitters)? Perhaps I'm missing something? What's your motivation for this approach in general?